cub/cub/device/dispatch/tuning/tuning_scan_by_key.cuh

```diff
@@ -587,543 +423,377 @@ template <class KeyT,
+                           default_reduce_by_key_delay_constructor_t<AccumT, int>>;
   };
 
-  struct DefaultTuning
+  template <CacheLoadModifier LoadModifier, typename DelayConstructurValueT>
+  struct DefaultPolicy
   {
-    static constexpr int NOMINAL_4B_ITEMS_PER_THREAD = 9;
-    static constexpr int ITEMS_PER_THREAD =
-      ((MaxInputBytes <= 8) ? 9 : Nominal4BItemsToItemsCombined(NOMINAL_4B_ITEMS_PER_THREAD, CombinedInputBytes));
+    static constexpr int nominal_4b_items_per_thread = 9;
+    static constexpr int items_per_thread =
+      max_input_bytes <= 8 ? 9 : Nominal4BItemsToItemsCombined(nominal_4b_items_per_thread, combined_input_bytes);
 
     using ScanByKeyPolicyT =
       AgentScanByKeyPolicy<256,
-                           ITEMS_PER_THREAD,
+                           items_per_thread,
                            BLOCK_LOAD_WARP_TRANSPOSE,
-                           LOAD_CA,
+                           LoadModifier,
                            BLOCK_SCAN_WARP_SCANS,
                            BLOCK_STORE_WARP_TRANSPOSE,
-                           detail::default_reduce_by_key_delay_constructor_t<AccumT, int>>;
+                           default_reduce_by_key_delay_constructor_t<DelayConstructurValueT, int>>;
   };
 
-  // SM520
   struct Policy520
-      : DefaultTuning
+      : DefaultPolicy<LOAD_CA, AccumT>
       , ChainedPolicy<520, Policy520, Policy350>
   {};
 
-  // SM800
+  // Use values from tuning if a specialization exists, otherwise pick the default
+  template <typename Tuning>
+  static auto select_agent_policy(int)
+    -> AgentScanByKeyPolicy<Tuning::threads,
+                            Tuning::items,
+                            Tuning::load_algorithm,
+                            LOAD_DEFAULT,
+                            BLOCK_SCAN_WARP_SCANS,
+                            Tuning::store_algorithm,
+                            typename Tuning::delay_constructor>;
+
+  template <typename Tuning>
+  // FIXME(bgruber): should we rather use `AccumT` instead of `ValueT` like the other default policies?
+  static auto select_agent_policy(long) -> typename DefaultPolicy<LOAD_DEFAULT, ValueT>::ScanByKeyPolicyT;
+
   struct Policy800 : ChainedPolicy<800, Policy800, Policy520>
   {
-    using tuning = detail::scan_by_key::sm80_tuning<KeyT, ValueT, detail::scan_by_key::is_primitive_op<ScanOpT>()>;
-
-    using ScanByKeyPolicyT =
-      AgentScanByKeyPolicy<tuning::threads,
-                           tuning::items,
-                           tuning::load_algorithm,
-                           LOAD_DEFAULT,
-                           BLOCK_SCAN_WARP_SCANS,
-                           tuning::store_algorithm,
-                           typename tuning::delay_constructor>;
+    // FIXME(bgruber): is `ValueT` correct here? Shouldn't we use `AccumT` instead?
+    using ScanByKeyPolicyT = decltype(select_agent_policy<sm80_tuning<key_t, ValueT, is_primitive_op<ScanOpT>()>>(0));
```

Comment:
@bernhardmgruber I think template parameter name is a copy-paste mistake. It should've been `ValueT`. On the usage side, it's correct and we should pass `ValueT`. `AccumT` is a derivative of the `KeyT` and operator. If we don't pass `ValueT`, there'll be no way of tuning an algorithm based on how much data we load from values iterator. My intuition is that value type should affect the performance more then accumulator type. Ideally, we would pass all of them (key, accum, and value types), but if I have to choose between accum and value, it's value type because I expect memory traffic from values to affect perf more than register pressure / compute.

Reply:
Thx a lot. I will clarify this here: #3143